shell/server/permissions.js

```diff
@@ -0,0 +1,212 @@
+
+function PermissionSet(array) {
+  // A wrapper around an array of booleans.
+
+  if (!array) {
+    this.array = [];
+  } else if (array instanceof Array) {
+    this.array = array;
+  } else {
+    throw new Error("don't know how to interpret as PermissionSet: " + array);
+  }
+}
+
+// Methods for mutating a PermissionSet by combining it with another PermissionSet.
+// These return a boolean indicating whether the operation had any effect.
+
+PermissionSet.prototype.add = function(other) {
+  var changed = false;
+  for (var ii = 0; ii < other.array.length; ++ii) {
+    var old = this.array[ii];
+    this.array[ii] = this.array[ii] || other.array[ii];
+    if (old != this.array[ii]) {
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+PermissionSet.prototype.remove = function(other) {
+  var changed = false;
+  for (var ii = 0; ii < other.array.length && ii < this.array.length; ++ii) {
+    var old = this.array[ii];
+    this.array[ii] = this.array[ii] && !other.array[ii];
+    if (old != this.array[ii]) {
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+PermissionSet.prototype.intersect = function(other) {
+  var changed = false;
+  for (var ii = 0; ii < this.array.length; ++ii) {
+    var old = this.array[ii];
+    this.array[ii] = this.array[ii] && other.array[ii];
+    if (old != this.array[ii]) {
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+function roleAssignmentPermissions(roleAssignment, viewInfo) {
+  var result = new PermissionSet([]);
+  if (!roleAssignment) {
+    return result;
+  }
+
+  if ("allAccess" in roleAssignment) {
+    var length = 0;
+    if (viewInfo.permissions) {
+      length = viewInfo.permissions.length;
+    }
+    var array = new Array(length);
+    for (var ii = 0; ii < array.length; ++ii) {
+      array[ii] = true;
+    }
+    result = new PermissionSet(array);
+  } else if ("roleId" in roleAssignment && viewInfo.roles && viewInfo.roles.length > 0) {
+    var roleDef = viewInfo.roles[roleAssignment.roleId];
+    if (roleDef) {
+      result = new PermissionSet(roleDef.permissions);
+    }
+  }
+
+  result.add(new PermissionSet(roleAssignment.addPermissionSet));
+  result.remove(new PermissionSet(roleAssignment.removePermissionSet));
+  return result;
+}
+
+function grainPermissionsInternal(grainId, userId, viewInfo) {
+  // Computes the permissions of the user by propagating permissions out from the grain's
+  // owner along role assignments until the graph is saturated.
+  //
+  // TODO(perf): This actually computes the permissions for *all* users. We should investigate ways
+  // to share results between calls. It may also make sense to cache some graph-connectivity
+  // information from the`mayOpenGrain()` computation to use here.
+
+  var grain = Grains.findOne(grainId);
+  var owner = grain.userId;
+
+  // userId -> PermissionSet
+  var permissionsMap = {}
+  permissionsMap[owner] = roleAssignmentPermissions({allAccess: null}, viewInfo);
+
+  var userStack = [owner];
```

Comment:
Instead of searching "forwards" from the owner (ultimately to all leaf nodes), what about searching "backwards" from the requesting user up to the owner? I think in most cases this would cut a large portion of the graph out of consideration.